language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,311 @@
+use crate::{counters::*, create_access_path, data_cache::StateViewCache};
+use anyhow::Result;
+use diem_state_view::StateView;
+use diem_types::{
+    account_address::AccountAddress,
+    account_config::{self, CurrencyInfoResource, RoleId},
+    transaction::{
+        GovernanceRole, SignatureCheckedTransaction, SignedTransaction, VMValidatorResult,
+    },
+    vm_status::{StatusCode, VMStatus},
+};
+use move_core_types::{
+    identifier::{IdentStr, Identifier},
+    move_resource::MoveStructType,
+    resolver::MoveResolver,
+};
+use move_vm_runtime::session::Session;
+
+use crate::logging::AdapterLogSchema;
+use diem_types::transaction::{Transaction, TransactionOutput, TransactionStatus, WriteSetPayload, TransactionPayload, TransactionArgument};
+use diem_types::access_path::AccessPath;
+use diem_logger::prelude::*;
+use rayon::prelude::*;
+use diem_types::write_set::WriteSet;
+use tracing::field::debug;
+use diem_types::block_metadata::BlockMetadata;
+use std::collections::HashSet;
+
+/// This trait describes the VM adapter's interface.
+pub trait VMAdapter {
+    /// Creates a new Session backed by the given storage.
+    /// TODO: this probably doesn't belong in this trait. We should be able to remove
+    /// this after refactoring Diem transaction executor code.
+    fn new_session<'r, R: MoveResolver>(&self, remote: &'r R) -> Session<'r, '_, R>;
+
+    /// Checks the signature of the given signed transaction and returns
+    /// `Ok(SignatureCheckedTransaction)` if the signature is valid.
+    fn check_signature(txn: SignedTransaction) -> Result<SignatureCheckedTransaction>;
+
+    /// Check if the transaction format is supported.
+    fn check_transaction_format(&self, txn: &SignedTransaction) -> Result<(), VMStatus>;
+
+    fn get_gas_price<S: MoveResolver>(&self, txn: &SignedTransaction, remote_cache: &S) -> Result<u64, VMStatus>;
+
+    /// Runs the prologue for the given transaction.
+    fn run_prologue<S: MoveResolver>(
+        &self,
+        session: &mut Session<S>,
+        transaction: &SignatureCheckedTransaction,
+        log_context: &AdapterLogSchema,
+    ) -> Result<(), VMStatus>;
+
+    fn should_restart_execution(output: &TransactionOutput) -> bool;
+
+    fn execute_single_transaction<S: MoveResolver + StateView>(
```

Comment:
StateView is a Diem-specific thing. You may want to get rid of StateView here, so that the adapter isn't specific to Diem.